### ISSUE TITLE
[4.1] Tests refactor

### ISSUE THREAD

**Issue by yguedidi** [facebook/facebook-php-sdk-v4#180]
Let's discut about how to refactor tests.
Here I show you my point of view, as I partially did in #175.

Three main points:
- Test public API, not implementation
- Don't use real Graph API calls
- A test helper for functional tests

## Test public API, not implementation
Many object method that have to be private (or protected) are public to "easily" test them.
But I think those method should be private, we shouldn't be able to call them from an instance.
Test should __only__ test the public API of a class, not its implementation.
We should can refactor a class (like split a big method in multiple one) without changing the tests.

## Don't use real Graph API calls
Tests of the SDK should only be __unit tests__. Making graph calls is useless, we can use fake data.
Graph calls could be useful only if we want to test the Graph API itself, but I think it's already fully tested by another team, and it's not the goal of the SDK.

## A test helper for functional tests
End user should be able to make functional tests using a test app. It's possible through a test helper.
Like any other helper, it will get a special access token for a given test app.
You can see how I implemented it in #175.

@gfosco, @SammyK any thoughts ? :)

**Comment by SammyK** [comment]
@yguedidi - Are we good to close this and revisit the testing stuff after 4.1? :)

**Comment by yguedidi** [comment]
Don't know.. Maybe we can simplify more the public API, and extension points like protected properties/methods, which will impact on how to write tests.

And strictly about tests, I think they should be fixed/cleaned/reworked before a release as big as 4.1 (kind of BDD point of view). But as we don't have a proper CI workflow (Travis-CI is not enabled), maybe we can work on tests after releasing..

As @gfosco want! :)

**Comment by SammyK** [comment]
Yeah, I hear you. At this point I think it's best to just get a release out and have a strict release cycle in place so that it's easier for users to upgrade with bite-sized BC's between each `MAJOR` release. And we won't have to feel like we have a "perfect product" for the 4.1/5.0 release. If we did, the next version of the SDK could be named [Facebook PHP SDK vDNF](http://en.wikipedia.org/wiki/Development_of_Duke_Nukem_Forever).

**Comment by gfosco** [comment]
Closing this.. if there are still points to discuss, please start a new issue.

**Comment by yguedidi** [comment]
This one depends on #447
